base-server.d.ts

```diff
@@ -0,0 +1,759 @@
+   */
+  redis?: string
+
+  /**
+   * Host to bind HTTP server to control Logux server.
+   * Default is `127.0.0.1`.
+   */
+  controlHost?: string
+
+  /**
+   * Port to control the server. Default is `31338`.
+   */
+  controlPort?: number
+
+  /**
+   * Password to control the server.
+   */
+  controlPassword?: string
+
+  /**
+   * Store to save log. Will be {@link @logux/core:MemoryStore}, by default.
+   */
+  store?: Store
+
+  /**
+   * Test time to test server.
+   */
+  time?: TestTime
+
+  /**
+   * Custom random ID to be used in node ID.
+   */
+  id?: string
+
+  /**
+   * Development or production server mode. By default,
+   * it will be taken from `NODE_ENV` environment variable.
+   * On empty `NODE_ENV` it will be `'development'`.
+   */
+  env?: 'production' | 'development'
+
+  /**
+   * Process ID, to display in reporter.
+   */
+  pid?: number
+
+  /**
+   * HTTP server to connect WebSocket server to it. Same as in `ws.Server`.
+   */
+  server?: HTTPServer
+
+  /**
+   * Port to bind server. It will create HTTP server manually to connect
+   * WebSocket server to it. Default is `31337`.
+   */
+  port?: number
+
+  /**
+   * IP-address to bind server. Default is `127.0.0.1`.
+   */
+  host?: string
+
+  /**
+   * SSL key or path to it. Path could be relative from server root.
+   * It is required in production mode, because WSS is highly recommended.
+   */
+  key?: string
+
+  /**
+   * SSL certificate or path to it. Path could be relative from server
+   * root. It is required in production mode, because WSS is highly
+   * recommended.
+   */
+  cert?: string
+
+  /**
+   * Function to show current server status.
+   */
+  reporter?: (event: string, payload: Object) => void
+}
+
+/**
+ * The authentication callback.
+ *
+ * @param userId User ID.
+ * @param credentials The client credentials.
+ * @param client Client object.
+ * @returns `true` if credentials was correct
+ */
+export type LoguxAuthenticator<Credentials = string> = (
+  userId: number | string | false,
+  credentials: Credentials | undefined,
+  server: ServerClient
+) => boolean | Promise<boolean>
+
+/**
+ * Check does user can do this action.
+ *
+ * @param ctx Information about node, who create this action.
+ * @param action The action data.
+ * @param meta The action metadata.
+ * @returns `true` if client are allowed to use this action.
+ */
+export type LoguxAuthorizer<
+  Action extends LoguxUserAction = LoguxUserAction
+> = (
+  ctx: Context,
+  action: Action,
+  meta: LoguxMeta
+) => boolean | Promise<boolean>
+
+/**
+ * Return object with keys for meta to resend action to other users.
+ *
+ * @param ctx Information about node, who create this action.
+ * @param action The action data.
+ * @param meta The action metadata.
+ * @returns Meta’s keys.
+ */
+export type LoguxResender<Action extends LoguxUserAction = LoguxUserAction> = (
+  ctx: Context,
+  action: Action,
+  meta: LoguxMeta
+) => Object | Promise<Object>
```

Comment:
I think we need to add `Resend` type with `{ users? string[], user?: string, … }`